bindings.c

```diff
@@ -80,6 +80,125 @@ struct file_info {
 	int cached;
 };
 
+/* The function of hash table.*/
+#define LOAD_SIZE 100 /*the size of hash_table */
+static int calc_hash(char *name)
+{
+	unsigned int hash = 0;
+	unsigned int x = 0;
+	/* ELFHash algorithm. */
+	while (*name) {
+		hash = (hash << 4) + *name++;
+		x = hash & 0xf0000000;
+		if (x != 0) {
+			hash ^= (x >> 24);
+			hash &= ~x;
+		}
+	}
+	return ((hash & 0x7fffffff) % LOAD_SIZE);
+}
+
+struct load_node {
+	char *cg;  /*cg */
+	unsigned long avenrun[3];		/* Load averages */
+	unsigned int run_pid;
+	unsigned int total_pid;
+	unsigned int last_pid;
+	int cfd; /* The file descriptor of the mounted cgroup */
+	struct  load_node *next;
+	struct  load_node **pre;
+};
+
+struct load_head {
+	/* The lock is about insert load_node and refresh load_node.To the first load_node of each hash bucket, insert and
+	refresh in this hash bucket is mutually exclusive. */
+	pthread_mutex_t lock;
+	/* The rdlock is about read loadavg and delete load_node.To each hash bucket, read and delete is
+	mutually exclusive. But at the same time, we allow paratactic read operation. This rdlock is at list level.*/
+	pthread_rwlock_t rdlock;
+	/* The rilock is about read loadavg and insert load_node.To the first load_node of each hash bucket, read and insert is
+	mutually exclusive. But at the same time, we allow paratactic read operation.*/
+	pthread_rwlock_t rilock;
+	struct load_node *next;
+};
+
+static struct load_head *load_hash[LOAD_SIZE]; /* hash table */
+
+static void init_load(void)
+{
+	int i;
+
+	for (i = 0; i < LOAD_SIZE; i++) {
+		load_hash[i] = (struct load_head *)malloc(sizeof(struct load_head));
+		load_hash[i]->next = NULL;
+		if (pthread_mutex_init(&load_hash[i]->lock, NULL) != 0)
+			goto out3;
+		if (pthread_rwlock_init(&load_hash[i]->rdlock, NULL) != 0)
+			goto out2;
+		if (pthread_rwlock_init(&load_hash[i]->rilock, NULL) != 0)
+			goto out1;
+	}
+	return;
+out1:
+	pthread_rwlock_destroy(&load_hash[i]->rdlock);
+out2:
+	pthread_mutex_destroy(&load_hash[i]->lock);
+out3:
+	free(load_hash[i]);
+	lxcfs_error("%s\n", "lock init error");
+	exit(1);
+}
+
+static void insert_node(struct load_node **n, int locate)
+{
+	pthread_mutex_lock(&load_hash[locate]->lock);
+	pthread_rwlock_wrlock(&load_hash[locate]->rilock);
+	struct load_node *f = load_hash[locate]->next;
+	load_hash[locate]->next = *n;
+
+	(*n)->pre = &(load_hash[locate]->next);
+	if (f)
+		f->pre = &((*n)->next);
+	(*n)->next = f;
+	pthread_mutex_unlock(&load_hash[locate]->lock);
+	pthread_rwlock_unlock(&load_hash[locate]->rilock);
+}
+/* Find spectial node. Not return NULL means success. */
+static struct load_node *locate_node(char *cg, int locate)
+{
+	struct load_node *f = NULL;
+	int i = 0;
+
+	pthread_rwlock_rdlock(&load_hash[locate]->rilock);
+	pthread_rwlock_rdlock(&load_hash[locate]->rdlock);
```

Comment:
Where do you unlock load_hash[locate]->rdlock ?

Reply:
This should be described more clearly. In **proc_loadavg_read**, I wrote a statement about unlock rdlock with comments. I will add when I unlock load_hash[locate]->rdlock here.